File: coordinator/api/model_manifest_contract.go
package api

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/eigeninference/d-inference/coordinator/store"
)

// These limits are the schema-v1 transport/structure contract shared with the
// Swift provider. Keep the values aligned with ModelManifestContract.
const (
	maxModelManifestBytes     int64 = 1 << 20
	maxModelManifestFileCount       = 16_384
)

func validateModelManifest(manifest *store.ModelManifest, modelID, version, r2Prefix string) error {
	if manifest == nil {
		return fmt.Errorf("manifest is empty")
	}
	if manifest.SchemaVersion != 1 {
		return fmt.Errorf("unsupported manifest schema_version %d", manifest.SchemaVersion)
	}
	if manifest.ModelID != modelID || manifest.Version != version || manifest.R2Prefix != r2Prefix {
		return fmt.Errorf("manifest fields do not match registration request")
	}
	if !isLowerSHA256Hex(manifest.AggregateSHA256) {
		return fmt.Errorf("manifest aggregate_sha256 must be 64 lowercase hex characters")
	}
	if manifest.TotalSizeBytes < 0 {
		return fmt.Errorf("manifest total_size_bytes must be nonnegative")
	}
	if manifest.FileCount < 1 {
		return fmt.Errorf("manifest must contain at least one file")
	}
	if manifest.FileCount > maxModelManifestFileCount {
		return fmt.Errorf("manifest file_count exceeds limit of %d", maxModelManifestFileCount)
	}
	if len(manifest.Files) > maxModelManifestFileCount {
		return fmt.Errorf("manifest files length exceeds limit of %d", maxModelManifestFileCount)
	}
	if manifest.FileCount != len(manifest.Files) {
		return fmt.Errorf("manifest file_count does not match files length")
	}

	totalSize, err := checkedManifestTotalSize(manifest.Files)
	if err != nil {
		return err
	}
	if totalSize != manifest.TotalSizeBytes {
		return fmt.Errorf("manifest total_size_bytes does not match files sum")
	}
	if aggregate := aggregateManifestFileHashes(manifest.Files); aggregate != manifest.AggregateSHA256 {
		return fmt.Errorf("manifest aggregate_sha256 does not match file hashes")
	}
	return nil
}

func checkedManifestTotalSize(files []store.ManifestFile) (int64, error) {
	var totalSize int64
	seenPaths := make(map[string]bool, len(files))
	for _, file := range files {
		if err := validateManifestFile(file); err != nil {
			return 0, err
		}
		pathKey := strings.ToLower(file.Path)
		if seenPaths[pathKey] {
			return 0, fmt.Errorf("manifest file path %q is duplicated", file.Path)
		}
		seenPaths[pathKey] = true
		if file.SizeBytes > math.MaxInt64-totalSize {
			return 0, fmt.Errorf("manifest file sizes overflow int64")
		}
		totalSize += file.SizeBytes
	}
	return totalSize, nil
}

func aggregateManifestFileHashes(files []store.ManifestFile) string {
	sorted := append([]store.ManifestFile(nil), files...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Path < sorted[j].Path })
	h := sha256.New()
	for _, file := range sorted {
		digest, err := hex.DecodeString(file.SHA256)
		if err != nil || len(digest) != sha256.Size {
			return ""
		}
		_, _ = h.Write(digest)
	}
	return hex.EncodeToString(h.Sum(nil))
}

func validateManifestFile(file store.ManifestFile) error {
	if !validManifestRelativePath(file.Path) {
		return fmt.Errorf("manifest file path %q is invalid", file.Path)
	}
	if file.SizeBytes < 0 {
		return fmt.Errorf("manifest file %q size_bytes must be nonnegative", file.Path)
	}
	if !isLowerSHA256Hex(file.SHA256) {
		return fmt.Errorf("manifest file %q sha256 must be 64 lowercase hex characters", file.Path)
	}
	return nil
}

func validManifestRelativePath(path string) bool {
	if path == "" || strings.HasPrefix(path, "/") || strings.Contains(path, "\\") {
		return false
	}
	for _, part := range strings.Split(path, "/") {
		if part == "" || part == "." || part == ".." {
			return false
		}
	}
	return true
}

func isLowerSHA256Hex(value string) bool {
	if len(value) != 64 {
		return false
	}
	for _, r := range value {
		if !(r >= '0' && r <= '9') && !(r >= 'a' && r <= 'f') {
			return false
		}
	}
	return true
}

File: provider-swift/Sources/ProviderCoreFoundation/ModelManifestContract.swift
import Foundation

/// Schema-v1 transport and structure limits for model manifests.
///
/// The coordinator mirrors these values when it accepts a registry publish.
/// Keeping the contract next to the wire types lets both coordinator-backed
/// and direct-CDN provider paths use exactly the same validation.
public enum ModelManifestContract {
    public static let maximumEncodedBytes = 1 * 1024 * 1024
    public static let maximumFileCount = 16_384
    public static let maximumStringBytes = 256 * 1024

    public enum ValidationError: Error, CustomStringConvertible, LocalizedError, Sendable {
        case encodedSizeExceeded(actual: Int)
        case unsupportedSchemaVersion(Int)
        case fileCountOutOfRange(Int)
        case fileArrayCountOutOfRange(Int)
        case fileCountMismatch(declared: Int, actual: Int)
        case fieldTooLarge
        case negativeTotalSize(Int64)
        case negativeFileSize(path: String, size: Int64)
        case duplicateFilePath(path: String)
        case totalSizeOverflow
        case totalSizeMismatch(declared: Int64, actual: Int64)

        public var description: String {
            switch self {
            case .encodedSizeExceeded(let actual):
                return "manifest response is \(actual) bytes; limit is \(ModelManifestContract.maximumEncodedBytes)"
            case .unsupportedSchemaVersion(let version):
                return "unsupported manifest schema_version \(version)"
            case .fileCountOutOfRange(let count):
                return "manifest file_count \(count) is outside 1...\(ModelManifestContract.maximumFileCount)"
            case .fileArrayCountOutOfRange(let count):
                return "manifest files count \(count) is outside 1...\(ModelManifestContract.maximumFileCount)"
            case .fileCountMismatch(let declared, let actual):
                return "manifest file_count \(declared) does not match files count \(actual)"
            case .fieldTooLarge:
                return "manifest field exceeds provider structural bound"
            case .negativeTotalSize(let size):
                return "manifest total_size_bytes \(size) must be nonnegative"
            case .negativeFileSize(let path, let size):
                return "manifest file \(path) size_bytes \(size) must be nonnegative"
            case .duplicateFilePath(let path):
                return "manifest file path \(path) is duplicated"
            case .totalSizeOverflow:
                return "manifest file sizes overflow Int64"
            case .totalSizeMismatch(let declared, let actual):
                return "manifest total_size_bytes \(declared) does not match files sum \(actual)"
            }
        }

        public var errorDescription: String? { description }
    }

    public static func validateEncodedByteCount(_ count: Int) throws {
        guard count <= maximumEncodedBytes else {
            throw ValidationError.encodedSizeExceeded(actual: count)
        }
    }

    public static func validate(_ manifest: ModelManifest) throws {
        guard manifest.schemaVersion == 1 else {
            throw ValidationError.unsupportedSchemaVersion(manifest.schemaVersion)
        }
        guard (1...maximumFileCount).contains(manifest.fileCount) else {
            throw ValidationError.fileCountOutOfRange(manifest.fileCount)
        }
        guard (1...maximumFileCount).contains(manifest.files.count) else {
            throw ValidationError.fileArrayCountOutOfRange(manifest.files.count)
        }
        guard manifest.fileCount == manifest.files.count else {
            throw ValidationError.fileCountMismatch(
                declared: manifest.fileCount,
                actual: manifest.files.count)
        }

        let topLevelStrings = [
            manifest.modelID,
            manifest.version,
            manifest.r2Prefix,
            manifest.aggregateSHA256,
        ]
        guard topLevelStrings.allSatisfy({ $0.utf8.count <= maximumStringBytes }),
              manifest.files.allSatisfy({ file in
                  file.path.utf8.count <= maximumStringBytes
                      && file.sha256.utf8.count <= maximumStringBytes
                      && file.role.utf8.count <= maximumStringBytes
              })
        else {
            throw ValidationError.fieldTooLarge
        }

        guard manifest.totalSizeBytes >= 0 else {
            throw ValidationError.negativeTotalSize(manifest.totalSizeBytes)
        }
        let total = try checkedTotalSize(manifest.files)
        guard total == manifest.totalSizeBytes else {
            throw ValidationError.totalSizeMismatch(
                declared: manifest.totalSizeBytes,
                actual: total)
        }
    }

    public static func checkedTotalSize(_ files: [ManifestFile]) throws -> Int64 {
        var total: Int64 = 0
        var seenPaths = Set<String>()
        for file in files {
            guard file.sizeBytes >= 0 else {
                throw ValidationError.negativeFileSize(path: file.path, size: file.sizeBytes)
            }
            // Match the coordinator's lowercase uniqueness key for download destinations.
            guard seenPaths.insert(file.path.lowercased()).inserted else {
                throw ValidationError.duplicateFilePath(path: file.path)
            }
            let (next, overflow) = total.addingReportingOverflow(file.sizeBytes)
            guard !overflow else {
                throw ValidationError.totalSizeOverflow
            }
            total = next
        }
        return total
    }
}
